pkgs/c_compiler/lib/src/tool/tool_resolver.dart

```diff
@@ -0,0 +1,249 @@
+      final file = File(LineSplitter.split(process.stdout).first);
+      final uri = File(await file.resolveSymbolicLinks()).uri;
+      return uri;
+    }
+    // The exit code for executable not being on the `PATH`.
+    assert(process.exitCode == 1);
+    return null;
+  }
+}
+
+class CliVersionResolver implements ToolResolver {
+  ToolResolver wrappedResolver;
+
+  CliVersionResolver({required this.wrappedResolver});
+
+  @override
+  Future<List<ToolInstance>> resolve({Logger? logger}) async {
+    final toolInstances = await wrappedResolver.resolve(logger: logger);
+    return [
+      for (final toolInstance in toolInstances)
+        await lookupVersion(toolInstance, logger: logger)
+    ];
+  }
+
+  static Future<ToolInstance> lookupVersion(
+    ToolInstance toolInstance, {
+    Logger? logger,
+  }) async {
+    if (toolInstance.version != null) return toolInstance;
+    logger?.finer('Looking up version with --version for $toolInstance.');
+    final version = await executableVersion(toolInstance.uri, logger: logger);
+    final result = toolInstance.copyWith(version: version);
+    logger?.fine('Found version for $result.');
+    return result;
+  }
+
+  static Future<Version> executableVersion(
+    Uri executable, {
+    String argument = '--version',
+    int expectedExitCode = 0,
+    Logger? logger,
+  }) async {
+    final process = await runProcess(
+      executable: executable,
+      arguments: [argument],
+      logger: logger,
+    );
+    if (process.exitCode != expectedExitCode) {
+      final executablePath = executable.toFilePath();
+      throw ToolError(
+          '`$executablePath $argument` returned unexpected exit code: '
+          '${process.exitCode}.');
+    }
+    return versionFromString(process.stdout)!;
+  }
+}
+
+class PathVersionResolver implements ToolResolver {
+  ToolResolver wrappedResolver;
+
+  PathVersionResolver({required this.wrappedResolver});
+
+  @override
+  Future<List<ToolInstance>> resolve({Logger? logger}) async {
+    final toolInstances = await wrappedResolver.resolve(logger: logger);
+
+    return [
+      for (final toolInstance in toolInstances) lookupVersion(toolInstance)
+    ];
+  }
+
+  static ToolInstance lookupVersion(ToolInstance toolInstance) {
+    if (toolInstance.version != null) {
+      return toolInstance;
+    }
+    return toolInstance.copyWith(
+      version: version(toolInstance.uri),
+    );
+  }
+
+  static Version? version(Uri uri) {
+    final versionString = uri.pathSegments.where((e) => e != '').last;
+    final version = versionFromString(versionString);
+    return version;
+  }
+}
+
+/// A resolver which invokes all [resolvers] tools.
+class ToolResolvers implements ToolResolver {
+  final List<ToolResolver> resolvers;
+
+  ToolResolvers(this.resolvers);
+
+  @override
+  Future<List<ToolInstance>> resolve({Logger? logger}) async => [
+        for (final resolver in resolvers)
+          ...await resolver.resolve(logger: logger)
```

Comment:
General comment applicable to here and some other places: You can consider `Future.wait`ing multiple futures instead of awaiting them individually.

Reply:
I was trying to limit concurrency as we're not using `package:pool` or anything to manage concurrency.

In a previous version I did have support for threading a `Pool` through, but I cut it out of this version. I might add it back in if doing everything serial is deemed too slow at some point.